src/wxMaxima.cpp

```diff
@@ -2640,8 +2640,8 @@ void wxMaxima::ReadVariables(wxString &data)
             }
             if(name == "gnuplot_command")
             {
-              m_gnuplotcommand = value;
-              wxLogMessage(wxString::Format(_("Gnuplot can be found at %s"),value.utf8_str()));
```

Comment:
This is the only place we can improve the thing in: Your heuristics for finding gnuplot is fine. But we shouldn't overwrite m_gnuplotCommand if we can execute the gnuplot maxima told us to use: The user might have two versions of gnuplot installed and might have customized maxima to use the only one that works.

The following piece of code from wxMaxima.cpp should do half of the task of verifying if the gnuplot maxima offers is somewhere in the search path:

    // Find gnuplot
    wxString gnuplot_binary;
    if(wxFileName(m_gnuplotcommand).IsAbsolute())
      gnuplot_binary = m_gnuplotcommand;
    else
    {
      wxPathList pathlist;
      pathlist.Add(wxFileName(wxStandardPaths::Get().GetExecutablePath()).GetPath());
      pathlist.Add(wxFileName(wxStandardPaths::Get().GetExecutablePath()).GetPath()+"/../");
      pathlist.Add(wxFileName(wxStandardPaths::Get().GetExecutablePath()).GetPath()+"/../gnuplot");
      pathlist.Add(wxFileName(wxStandardPaths::Get().GetExecutablePath()).GetPath()+"/../gnuplot/bin");
      pathlist.AddEnvList(wxT("PATH"));
      gnuplot_binary = pathlist.FindAbsoluteValidPath(m_gnuplotcommand);
      if(gnuplot_binary == wxEmptyString)
        gnuplot_binary = pathlist.FindAbsoluteValidPath(m_gnuplotcommand + wxT(".exe"));
      if(gnuplot_binary == wxEmptyString)
        gnuplot_binary = pathlist.FindAbsoluteValidPath(m_gnuplotcommand + wxT(".exe"));
      if(gnuplot_binary == wxEmptyString)
        gnuplot_binary = m_gnuplotcommand;
    }

In theory if you replace m_gnuplotCommand by Dirstructure::GnuplotDefaultLocation(value) in this place it should be early enough.

...and as a 2nd idea: If wxMaxima knows where gnuplot resides that doesn't help if one wants to instruct maxima to plot a new diagram: Maxima needs to know where gnuplot resides, too, in order to be able to do that. And if maxima knows where gnuplot can be found it will tell us the right gnuplot location. If maxima currently sends us the wrong info: Would you be so kind and tell them that they need to fix this? As a workaround until that is effective we could add a maxima command that corrects the path to gnuplot to m_configCommands. Or does maxima know the correct path to gnuplot because it is run in a shell? And is us not knowing where gnuplot resides a problem only wxMaxima has, because it is a gui application? In this case I would propose that we use the gnuplot maxima tells us to use if we can find it (the user might have customized it), and use the gnuplot we have found, otherwise.

Reply:
Fine, i can change back changing m_gnuplotcommand. I did this mostly for performance reasons, but since we don't search paths recursively, it should be fine.

You probably have seen that I also changed the command which (on OSX) injects a command to set the gnuplot path for maxima. This was there before, but used a different heuristic to find gnuplot, which in my view wasn't a good idea.

I didn't find a place where you retrieve the gnuplot command from maxima. Can you point me to it?

Reply:
If maxima informs us about the value of a variable 

     wxMaxima::ReadVariables(wxString &data)

is called. The exact code location is:

            if(name == "gnuplot_command")
            {
              m_gnuplotcommand = value;

Having the gnuplot command configurable by maxima is especially usable on MS Windows: For windows there are two versions of gnuplot that each come with different bugs:

  * gnuplot.exe plots embedded and other plots fine if run from a terminal (and only shows any gnuplot window if the terminal actually opens up in a separate, and complete useless and empty window), but some users want to be able to enable gnuplot's command console. And gnuplot.exe opens the command console in a hidden, and inaccessible window.
  * wgnuplot.exe works just fine: It pops up a window if you want to, is able to open its command console. Bnd even if it plots an embedded plot for the worksheet or an animation frame and therefore doesn't open a window it steals the mouse focus which makes working with plots a nuisance.

I am always thinking of telling maxima to report the value of gnuplot_command every time maxima goes busy, which would allow to change wxMaxima's m_gnuplot_command from the worksheet.